### PR TITLE
Refactor: Rebuild homepage layout with Mustache

### PR DESCRIPTION
**Pull request by Retr0ve**
- Rebuilt with Mustache
- Run `yarn build` to build, content will be generated at `./_site` folder

### CODE REVIEW

File: build.js
Comment (laurent22) [inline build.js:6]: Everything under the Joplin org needs to use tabs, not spaces.

File: build.js
Comment (laurent22) [inline build.js:21]: why is that async when all your code is synchronous.

File: data/allPossibleCategories.js
Comment (laurent22) [inline data/allPossibleCategories.js:2]: Can't have this duplicate list here

File: utils/generateHTMLs.js
Comment (laurent22) [inline utils/generateHTMLs.js:1]: Why is this all JS code anyway? That should all be implemented in TypeScript.

File: build.js
Comment (Retr0ve) [reply build.js]: It is because I truncate this code from a larger piece, and it contains asynchronous methods. I've removed it now, but it might come back in the future.

File: utils/generateHTMLs.js
Comment (Retr0ve) [reply utils/generateHTMLs.js]: I've turned my code to ts now.

File: data/allPossibleCategories.js
Comment (Retr0ve) [reply data/allPossibleCategories.js]: I've raise a PR in the main codebase to maintain this list in @joplin/lib, I'll include @joplin/lib to get this list after that PR is merged.

File: build.js
Comment (Retr0ve) [reply build.js]: Done

File: pages/index.mustache
Comment (laurent22) [inline pages/index.mustache:158]: Please pay attention to indentation and mix of tabs and spaces. This is basic thing that should be correct. You can also install a linter to take care of that automatically

Comment (Retr0ve) [reply pages/index.mustache:151]: I've added esllint and prettier to format code, it should all be tabs now.

File: data/plugins.ts
Comment (Retr0ve) [reply data/plugins.ts]: fixed

File: data/plugins.ts
Comment (Retr0ve) [reply data/plugins.ts]: done

**Comment by laurent22** [comment]
Is there any reason why all this work was done in a different repo? The current website build files are all in the monorepo and for a good reason - so that they can access various lib files and tools and take advantage of the existing TypeScript and eslint setup.

Here you're already having trouble accessing joplin/lib and you are copying and pasting images and icons, recreating the eslint and TS config, etc. We can't have this - maybe the most important part of this project is to automate as much as possible and make the life of future maintainers easier.

The image creation for example needs to be automated using the generate-image from the main repo.

I'm not sure what to do about all this. The fact that you left this PR forgotten here for 30 days doesn't help.

**Comment by Retr0ve** [comment]
Hi Laurent,
Thanks for your review.

I don't know if it's possible to create several github pages in one repo. And also, my mentor create this repo for me, so I assume you want me to work in this repo. If this should be a part of the main repo, I think it will be better to notify me beforehand. As a solution, I suggest that we can still working on this repo and gradually move it into the main repo.

Besides, I'm having trouble accessing joplin/lib because my PR is still open. I'm confusing that why you said it was merged since it is not.

As a fact, I didn't leave this PR here for more than a month, I asked several times whether if there is someone reviewing it and I got no positive feedback. And you explicitly said that don't tag you, so I got nowhere else to turn to. 

I am suggesting that we can work with current code to make it work at least, then gradually refine it by replacing the code such as pasting images to libs you mentioned and merging it into the main repo.

**Comment by laurent22** [comment]
> Besides, I'm having trouble accessing joplin/lib because my PR is still open. I'm confusing that why you said it was merged since it is not.

Yes but even if we merge, we aren't going to be publishing the package immediately. Usually we do that at the end of a release cycle.

> I asked several times whether if there is someone reviewing it and I got no positive feedback

Not sure where you asked several times since I didn't see a single message from you.

> I am suggesting that we can work with current code to make it work at least, then gradually refine it by replacing the code such as pasting images to libs you mentioned and merging it into the main repo.

Well, yes I guess, at this point that's what will have to be done.

**Comment by Retr0ve** [comment]
> Yes but even if we merge, we aren't going to be publishing the package immediately. Usually we do that at the end of a release cycle.

Thats why I'm suggesting to do it in the next step.



> Not sure where you asked several times since I didn't see a single message from you.

It's in the forum DM. I've asked my mentors about it, it's until later I realize that they are not able to review my PR.

File: pages/index.mustache
Comment (Retr0ve) [reply pages/index.mustache]: Done

**Comment by laurent22** [comment]
Code seems ok but it's missing unit tests.

**Comment by Retr0ve** [comment]
OK. Should I raise a separate PR or just add them to this one?

**Comment by laurent22** [comment]
Yes, ideally tests should be part of the pull request. They are used to prove the code actually works, and often when writing these tests you need to refactor and improve the code.

File: data/allPossibleCategories.ts
Comment (personalizedrefrigerator) [inline data/allPossibleCategories.ts:35]: Consider removing this commented-out code (or adding a comment to explain why it's commented out).

File: data/allPossibleCategories.ts
Comment (personalizedrefrigerator) [inline data/allPossibleCategories.ts:26]:
```suggestion
			name: category.toLowerCase().replace(/\s/g, '-'),
```
I think matching all space characters with `\s` makes more sense here. Because we already know the category names, feel free to ignore this change!

File: pages/index.mustache
Comment (personalizedrefrigerator) [inline pages/index.mustache:49]:
```suggestion
				<p class="text-muted subtitle mb-3">Joplin has many plugins provided by the community which can be helpful
```
Feel free to ignore this suggestion!

**Review by personalizedrefrigerator — COMMENTED** [review]
Nice work so far!

The first time I run `yarn build`, I get the following error:
```console
tmpuser:/tmp/build/website-plugin-discovery$ yarn build
yarn run v1.22.19
$ tsc && node build.js
Error: ENOENT: no such file or directory, stat './site'
    at Object.statSync (node:fs:1596:3)
    at __node_internal_ (node:internal/fs/utils:804:8)
    at Object.rmSync (node:fs:1271:13)
    at clearBuildPath (/tmp/build/website-plugin-discovery/build.js:13:28)
    at /tmp/build/website-plugin-discovery/build.js:57:5
    at Object.<anonymous> (/tmp/build/website-plugin-discovery/build.js:62:3)
    at Module._compile (node:internal/modules/cjs/loader:1119:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1173:10)
    at Module.load (node:internal/modules/cjs/loader:997:32)
    at Module._load (node:internal/modules/cjs/loader:838:12) {
  errno: -2,
  syscall: 'stat',
  code: 'ENOENT',
  path: './site'
}
Rendering /tmp/build/website-plugin-discovery/site/index.html
(node:10457) ExperimentalWarning: The Fetch API is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Done in 2.35s.

```
Running `yarn build` again builds without warnings/errors.

**Comment by personalizedrefrigerator** [comment]
When I run this locally, the website lists no plugins:

What should I do to generate the list of plugins? How should I be serving the webpage?

**Comment by Retr0ve** [comment]
> When I run this locally, the website lists no plugins:  What should I do to generate the list of plugins? How should I be serving the webpage?

Hi, thanks for your review! This PR does not contain the code to generate plugin list. I'll create another PR as soon as this one gets merged. As for the bugs you mentioned, I'll fix them soon.

File: data/allPossibleCategories.ts
Comment (Retr0ve) [reply data/allPossibleCategories.ts:35]: these codes are prepared for pulling this list from joplin/lib. Maybe just leave them here for now, I'll replace them when lib v2.9 is released.

**Review by personalizedrefrigerator — APPROVED** [review]

**Pull request merged**
